File: src/main/java/com/laamella/code_state_machine/util/DslStateMachineBuilder.java
package com.laamella.code_state_machine.util;

import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.laamella.code_state_machine.Action;
import com.laamella.code_state_machine.Precondition;
import com.laamella.code_state_machine.StateMachine;

/**
 * A pretty "DSL" builder for a state machine.
 */
public class DslStateMachineBuilder<T, E> {
	private static final Logger log = LoggerFactory.getLogger(StateMachine.class);

	private Set<T> sourceStates = new HashSet<T>();
	private Precondition<E> storedPrecondition;
	private Action<E> storedAction;
	private final StateMachine.Builder<T, E> builder = new StateMachine.Builder<T, E>();

	public DslStateMachineBuilder() {
		resetStoredTransitionVariables();
	}

	public StateMachine<T, E> buildMachine() {
		return builder.build();
	}

	@SuppressWarnings("unchecked")
	public DslStateMachineBuilder<T, E> state(final T state) {
		return states(state);
	}

	public DslStateMachineBuilder<T, E> states(final T... states) {
		sourceStates = new HashSet<T>(Arrays.asList(states));
		return this;
	}

	public DslStateMachineBuilder<T, E> except(final T... states) {
		for (final T state : states) {
			sourceStates.remove(state);
		}
		return this;
	}

	public DslStateMachineBuilder<T, E> onExit(final Action<E> action) {
		for (final T sourceState : sourceStates) {
			builder.setExitAction(sourceState, action);
		}
		return this;
	}

	public DslStateMachineBuilder<T, E> onEntry(final Action<E> action) {
		for (final T sourceState : sourceStates) {
			log.debug("Create entry action for {} ({})", sourceState, action);
			builder.setEntryAction(sourceState, action);
		}
		return this;
	}

	public DslStateMachineBuilder<T, E> transition(final T destinationState, final Precondition<E> precondition,
			final Action<E> action) {
		for (final T sourceState : sourceStates) {
			builder.addTransition(new BasicTransition<T, E>(sourceState, destinationState, precondition, action));
		}
		resetStoredTransitionVariables();
		return this;
	}

	private void resetStoredTransitionVariables() {
		storedAction = nothing();
		storedPrecondition = always();
	}

	public DslStateMachineBuilder<T, E> when(final Precondition<E> precondition) {
		storedPrecondition = precondition;
		return this;
	}

	public DslStateMachineBuilder<T, E> when(final E... events) {
		storedPrecondition = is(events);
		return this;
	}

	public DslStateMachineBuilder<T, E> action(final Action<E> action) {
		storedAction = action;
		return this;
	}

	public DslStateMachineBuilder<T, E> then(final T destinationState) {
		return transition(destinationState, storedPrecondition, storedAction);
	}

	public Precondition<E> always() {
		return new AlwaysPrecondition<T, E>();
	}

	public Precondition<E> after(final long milliseconds) {
		return new AfterPrecondition<T, E>(milliseconds);
	}

	public Precondition<E> is(final E... events) {
		assert events != null;
		assert events.length != 0;

		if (events.length == 1) {
			final E singleEvent = events[0];
			return new SingleEventMatchPrecondition<T, E>(singleEvent);
		}

		return new MultiEventMatchPrecondition<T, E>(events);
	}

	public Action<E> nothing() {
		return new NoAction<T, E>();
	}

	public DslStateMachineBuilder<T, E> isEndState() {
		for (final T state : sourceStates) {
			builder.addEndState(state);
		}
		return this;
	}

	public DslStateMachineBuilder<T, E> isStartState() {
		for (final T state : sourceStates) {
			builder.addStartState(state);
		}
		return this;
	}

	public DslStateMachineBuilder<T, E> areEndStates() {
		return isEndState();
	}

	public DslStateMachineBuilder<T, E> areStartStates() {
		return isStartState();
	}
}

File: scala/src/main/scala/com/laamella/code_state_machine/builder/DslStateMachineBuilder.scala
package com.laamella.code_state_machine.builder

import com.laamella.code_state_machine._
import com.laamella.code_state_machine.action.LogAction
import com.laamella.code_state_machine.condition.{AfterCondition, AlwaysCondition, MultiEventMatchCondition, NeverCondition, SingleEventMatchCondition, StatesActiveCondition, StatesInactiveCondition}
import org.slf4j.LoggerFactory

import scala.collection.mutable

/**
 * A pretty "DSL" builder for a state machine.
 */
abstract class DslStateMachineBuilder[T, E, P <: Ordered[P]](defaultPriority: P) extends StateMachineBuilder[T, E, P] {
  private val log = LoggerFactory.getLogger("DslStateMachineBuilder")

  class DefiningState(sourceStates: mutable.Set[T], internals: StateMachine[T, E, P]#Internals) {
    //		private Set<T> sourceStates = new HashSet<T>();
    //		private final StateMachine<T, E, P>.Internals internals;
    //
    //		public DefiningState() {
    //			this.sourceStates = sourceStates;
    //			this.internals = internals;
    //		}

    def except(states: T*): DefiningState = {
      for (state <- states) {
        sourceStates.remove(state)
      }
      this
    }

    def onExit(action: Action*): DefiningState = {
      for (sourceState <- sourceStates) {
        internals.addExitActions(sourceState, action)
      }
      this
    }

    def onEntry(action: Action*): DefiningState = {
      for (sourceState <- sourceStates) {
        log.debug(s"Create entry action for $sourceState $action")
        internals.addEntryActions(sourceState, action);
      }
      this
    }

    def isAnEndState(): DefiningState = {
      for (state <- sourceStates) {
        internals.addEndState(state)
      }
      this
    }

    def isAStartState(): DefiningState = {
      for (state <- sourceStates) {
        internals.addStartState(state)
      }
      this
    }

    def areEndStates(): DefiningState = {
      isAnEndState()
    }

    def areStartStates(): DefiningState = isAStartState()

    def whenConditions(condition: Condition[E]*): DefiningTransition = {
      //	TODO		assert condition != null;
      new DefiningTransition(sourceStates, new Conditions[E](condition: _*), internals)
    }

    def whenEvents(events: E*): DefiningTransition = {
      new DefiningTransition(sourceStates, is(events:_*), internals)
    }

  }

  class DefiningTransition(sourceStates: mutable.Set[T], conditions: Conditions[E], internals: StateMachine[T, E, P]#Internals) {
    private val actions = new Actions()
    private var priority = defaultPriority

    //		public DefiningTransition() {
    //			this.sourceStates = sourceStates;
    //			this.conditions = conditions;
    //			this.internals = internals;
    //		}

    def action(action: Action): DefiningTransition = {
      //	TODO		assert action != null;
      actions.add(action)
      this
    }

    def then(destinationState: T): DefiningState = {
      //	TODO		assert destinationState != null;
      transition(destinationState, conditions, priority, actions)
    }

    def transition(destinationState: T, storedConditions2: Conditions[E], priority: P, actions: Actions): DefiningState = {
      this.actions.add(actions)
      for (sourceState <- sourceStates) {
        internals.addTransition(new Transition[T, E, P](sourceState, destinationState, storedConditions2, priority, this.actions))
      }
      new DefiningState(sourceStates, internals)
    }

    def transition(destinationState: T, condition: Condition[E], priority: P, actions: Action*): DefiningState = {
      transition(destinationState, new Conditions[E](condition), priority, new Actions(actions: _*))
    }

    def withPrio(priority: P): DefiningTransition = {
      //TODO			assert priority != null;
      this.priority = priority
      this
    }
  }

  private var machine: StateMachine[T, E, P] = _

  override def build(newMachine: StateMachine[T, E, P]): StateMachine[T, E, P] = {
    //	TODO	assert newMachine != null;
    machine = newMachine
    executeBuildInstructions()
    machine
  }

  // TODO see if we can change this pattern to get rid of the var machine
  protected def executeBuildInstructions(): Unit

  override def build(): StateMachine[T, E, P] = {
    build(new StateMachine[T, E, P]())
  }

  def state(state: T): DefiningState = {
    //TODO		assert state != null;
    states(state)
  }

  def states(states: T*): DefiningState = {
    val m = machine;
    new DefiningState(mutable.HashSet[T](states: _*), new m.Internals())
  }

  def active(statesThatMustBeActive: T*): Condition[E] = {
    new StatesActiveCondition[T, E, P](machine, statesThatMustBeActive: _*)
  }

  def inactive(statesThatMustBeInactive: T*): Condition[E] = {
    new StatesInactiveCondition[T, E, P](machine, statesThatMustBeInactive: _*)
  }

  def always() = new AlwaysCondition[E]()



  def never() = new NeverCondition[E]()

  def after(milliseconds: Long) = new AfterCondition[E](milliseconds)

  def is(events: E*): Conditions[E] = {
    //TODO					assert events != null;
    //					assert events.length != 0;
    if (events.length == 1) {
      val singleEvent = events(0)
      return new Conditions[E](new SingleEventMatchCondition[E](singleEvent))
    }
    new Conditions[E](new MultiEventMatchCondition[E](events: _*))
  }

  def log(logText: String): Action = {
    //TODO	assert logText != null;
    new LogAction(logText)
  }

}
